Core/Src/main.c

```diff
@@ -26,5 +42,26 @@ int main()
 void UART1_RxCallback(void) 
 {
   GPIO_TogglePin(LED_GREEN_Port, LED_GREEN_Pin);
-  UART1_ReceiveIT(buf, 1);
+  UART1_ReceiveIT(bufUART, 1);
+
+  if(bufUART[0] == 'd')
+  {
+    DS3231_Date_t date;
+    date.Year = 21;
+    date.Month = 1;
+    date.Date = 14;
+    date.Day = 3;
+
+    DS3231_SetDate(&DS3231, &date);
```

Comment:
You must not perform any peripheral related operations inside the interrupt.
The only thing you can do here is to set the flag that the data is received